File: programs/integer-replacement.go
// +build integerReplacement

package main

import "fmt"

func integerReplacement(n int) int {
	if n <= 3 {
		return n - 1
	} else {
		if n%2 == 0 {
			return integerReplacement(n/2) + 1
		} else {
			sub1 := integerReplacement(n - 1)
			plus1 := integerReplacement(n + 1)
			if sub1 > plus1 {
				return plus1 + 1
			} else {
				return sub1 + 1
			}
		}
	}
}

func main() {
	fmt.Println(integerReplacement(1))
	fmt.Println(integerReplacement(2))
	fmt.Println(integerReplacement(3))
	fmt.Println(integerReplacement(4))
	fmt.Println(integerReplacement(7))
	fmt.Println(integerReplacement(8))
}

File: programs/integer-replacement.js
var integerReplacement = function(n) {
  if (n <= 3) {
    return n - 1;
  } else {
    if (0 === n % 2) {
      return integerReplacement(n / 2) + 1;
    } else {
      return Math.min(integerReplacement(n - 1) + 1, integerReplacement(n + 1) + 1);
    }
  }
};

console.log(integerReplacement(1));
console.log(integerReplacement(2));
console.log(integerReplacement(3));
console.log(integerReplacement(4));
console.log(integerReplacement(7));
console.log(integerReplacement(8));